presto-main/src/test/java/io/prestosql/sql/query/TestRowFilter.java

```diff
@@ -0,0 +1,213 @@
+                    new ViewExpression(USER, Optional.empty(), Optional.empty(), "orderkey < 10"));
+            assertions.assertQuery("SELECT count(*) FROM orders", "VALUES BIGINT '7'");
+        });
+
+        assertions.executeExclusively(() -> {
+            accessControl.reset();
+            accessControl.rowFilter(
+                    new QualifiedObjectName(CATALOG, "tiny", "orders"),
+                    USER,
+                    new ViewExpression(USER, Optional.empty(), Optional.empty(), "NULL"));
+            assertions.assertQuery("SELECT count(*) FROM orders", "VALUES BIGINT '0'");
+        });
+    }
+
+    @Test
+    public void testCorrelatedSubquery()
+    {
+        assertions.executeExclusively(() -> {
+            accessControl.reset();
+            accessControl.rowFilter(
+                    new QualifiedObjectName(CATALOG, "tiny", "orders"),
+                    USER,
+                    new ViewExpression(USER, Optional.of(CATALOG), Optional.of("tiny"), "EXISTS (SELECT 1 FROM nation WHERE nationkey = orderkey)"));
+            assertions.assertQuery("SELECT count(*) FROM orders", "VALUES BIGINT '7'");
+        });
+    }
+
+    @Test
+    public void testRecursion()
+    {
+        assertions.executeExclusively(() -> {
+            accessControl.reset();
+            accessControl.rowFilter(
+                    new QualifiedObjectName(CATALOG, "tiny", "orders"),
+                    USER,
+                    new ViewExpression(USER, Optional.of(CATALOG), Optional.of("tiny"), "orderkey IN (SELECT orderkey FROM orders)"));
+
+            assertions.assertFails("SELECT count(*) FROM orders", ".*\\QRow filter for 'local.tiny.orders' is recursive\\E.*");
+        });
+
+        // different reference style to same table
+        assertions.executeExclusively(() -> {
+            accessControl.reset();
+            accessControl.rowFilter(
+                    new QualifiedObjectName(CATALOG, "tiny", "orders"),
+                    USER,
+                    new ViewExpression(USER, Optional.of(CATALOG), Optional.of("tiny"), "orderkey IN (SELECT local.tiny.orderkey FROM orders)"));
+            assertions.assertFails("SELECT count(*) FROM orders", ".*\\QRow filter for 'local.tiny.orders' is recursive\\E.*");
+        });
+    }
+
+    @Test
+    public void testOtherSchema()
+    {
+        assertions.executeExclusively(() -> {
+            accessControl.reset();
+            accessControl.rowFilter(
+                    new QualifiedObjectName(CATALOG, "tiny", "orders"),
+                    USER,
+                    new ViewExpression(USER, Optional.of(CATALOG), Optional.of("sf1"), "(SELECT count(*) FROM customer) = 150000")); // Filter is TRUE only if evaluating against sf1.customer
+            assertions.assertQuery("SELECT count(*) FROM orders", "VALUES BIGINT '15000'");
+        });
+    }
+
+    @Test
+    public void testDifferentIdentity()
+    {
+        assertions.executeExclusively(() -> {
```

Comment:
As I understand it will use `RowFilterKey` from `TestingAccessControlManager`. It would be nice to have a test that would require using different keys in analysis based on identity.

Reply:
Not sure I understand. This test does that -- it applies to different filters to two different instances of the `orders` table based on the identity being used to evaluate the tables.

Reply:
Unfortunately, anything involving views would require a connector that supports them AND supports filters/masks. We don’t have any.